Assignment2/Assignment 2.py

```diff
@@ -0,0 +1,375 @@
+
+# Plot the prediction results
+plt.scatter(valid_x, valid_y, s=10)
+plt.plot(valid_x, valid_y_poly_pred, color='r')
+plt.show()
+
+
+# ## Question 1: Polynomial Regression Using Sci-Kit Learn
+# 
+# Now it is your turn! Following the same format as above, implement a 10-degree polynomial regression model on the training data and plot your results. Use your model to predict the output of the validation set and calculate the mean square error. Report and plot the results. 
+
+# In[7]:
+
+
+### Fit a 10-degree polynomial using Sci-Kit Learn
+# Create polynomial features with degree 10
+polynomial_features = PolynomialFeatures(degree=10)
+x_poly = polynomial_features.fit_transform(train_x)
+
+# Apply linear regression
+model = LinearRegression()
+model.fit(x_poly, train_y)
+y_poly_pred = model.predict(x_poly)
+
+### Plot curve on the training data set
+plt.scatter(train_x, train_y, s=10)
+plt.plot(train_x, y_poly_pred, color='r')
+plt.show()
+
+### Use model to predict output of validation set
+valid_y_poly_pred = model.predict(polynomial_features.fit_transform(valid_x))
+
+### Calculate the RMSE. Report and plot the curve on the validation set.
+# Calculate root mean squared error
+mse = mean_squared_error(valid_y, valid_y_poly_pred)
+print("Mean Squared Error: {}".format(mse))
+
+# Plot the prediction results
+plt.scatter(valid_x, valid_y, s=10)
+plt.plot(valid_x, valid_y_poly_pred, color='r')
+plt.show()
+
+
+# #### Did the root mean squared error go up or down as compared to the 2-degree polynomial curve? Why do you think this is the case?
+# 
+# It went down compared to the 2-degree polynomial curve. A 10 degree polynomial is a better fit for the data.
+
+# Now repeat the above for a 20-degree polynomial regression model.
+
+# In[8]:
+
+
+### Fit a 10-degree polynomial using Sci-Kit Learn
+# Create polynomial features with degree 10
+polynomial_features = PolynomialFeatures(degree=20)
+x_poly = polynomial_features.fit_transform(train_x)
+
+# Apply linear regression
+model = LinearRegression()
+model.fit(x_poly, train_y)
+y_poly_pred = model.predict(x_poly)
+
+### Plot curve on the training data set
+plt.scatter(train_x, train_y, s=10)
+plt.plot(train_x, y_poly_pred, color='r')
+plt.show()
+
+### Use model to predict output of validation set
+valid_y_poly_pred = model.predict(polynomial_features.fit_transform(valid_x))
+
+### Calculate the RMSE. Report and plot the curve on the validation set.
+# Calculate root mean squared error
+mse = mean_squared_error(valid_y, valid_y_poly_pred)
+print("Mean Squared Error: {}".format(mse))
+
+# Plot the prediction results
+plt.scatter(valid_x, valid_y, s=10)
+plt.plot(valid_x, valid_y_poly_pred, color='r')
+plt.show()
+
+
+# #### How does the mean square error compare to the previous two models? Why do you think this is the case?
+# 
+# It went up. 20 degrees is too high, the model is learning the gaussian noise in the training data and performs poorly on test data.
```

Comment:
Yes! but also because the 20 degree polynomial overfits to the data, so captures the noise. 
Whereas the 2 degree polynomial was excessively simple, and generalizes too much (underfits the data)